### PR TITLE
Stickiness cookie name

### LINKED ISSUE #2175: Feature Request: Configurable sticky cookie name
### Do you want to request a *feature* or report a *bug*?

*feature*

This feature request is related to https://github.com/containous/traefik/issues/1574 and https://github.com/containous/traefik/pull/1716, https://github.com/containous/traefik/pull/1712 already had the perfect solution but was dropped.

For heavy traffic applications like ours it would be very good if the LB cookie name would be configurable, the old implementation already had a VERY long cookie name, now it can become even longer, what we see with 1.4RC3 is `_TRAEFIK_BACKEND_backend-adserver-production` at ~800M Requests per day, this cookie name alone accounts for ~35GB of traffic.

It would be great to have a) the backend addition https://github.com/containous/traefik/pull/1716#issuecomment-307330224 optional and b) to make the cookie name configurable, we are very careful with cookies, 99% of our users only have a cookie i=someid, so it would be good if we could use `t=` or `_=` for traefik.

### PR DESCRIPTION
**Pull request by ldez**
### Description

- add an option to configure manually cookie name (ensures compatibility).
- use a hash as cookie name by default.

Fixes #2175

### CODE REVIEW

File: types/types.go [hunk 1]
```diff
@@ -33,8 +33,14 @@ type MaxConn struct {
 
 // LoadBalancer holds load balancing configuration.
 type LoadBalancer struct {
-	Method string `json:"method,omitempty"`
-	Sticky bool   `json:"sticky,omitempty"`
+	Method     string      `json:"method,omitempty"`
+	Sticky     bool        `json:"sticky,omitempty"` // Deprecated
```
Comment (emilevauge) [inline types/types.go]: Could you complete to stick with the convention?
`// Deprecated: use Stickiness instead`

File: provider/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -247,8 +247,14 @@ func (p *Provider) loadIngresses(k8sClient Client) (*types.Configuration, error)
 					templateObjects.Backends[r.Host+pa.Path].LoadBalancer.Method = "drr"
 				}
 
-				if service.Annotations[types.LabelBackendLoadbalancerSticky] == "true" {
-					templateObjects.Backends[r.Host+pa.Path].LoadBalancer.Sticky = true
+				if len(service.Annotations[types.LabelBackendLoadbalancerSticky]) != 0 {
```
Comment (emilevauge) [inline provider/kubernetes/kubernetes.go]: `len(service.Annotations[types.LabelBackendLoadbalancerSticky]) > 0` would be even better :nerd_face:

File: provider/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -247,8 +247,14 @@ func (p *Provider) loadIngresses(k8sClient Client) (*types.Configuration, error)
...
+					log.Warn("Deprecated configuration found: %s. Please use %s.", types.LabelBackendLoadbalancerSticky, types.LabelBackendLoadbalancerStickiness)
+				}
+
+				if service.Annotations[types.LabelBackendLoadbalancerSticky] == "true" || service.Annotations[types.LabelBackendLoadbalancerStickiness] == "true" {
+					templateObjects.Backends[r.Host+pa.Path].LoadBalancer.Stickiness = &types.Stickiness{
+						CookieName: cookie.GenerateName(r.Host + pa.Path),
```
Comment (emilevauge) [inline provider/kubernetes/kubernetes.go]: Why not choosing the cookie name generation from hash here ? :thinking:

File: provider/consul/consul_catalog.go [hunk 4]
```diff
@@ -397,6 +393,26 @@ func (p *CatalogProvider) getBasicAuth(tags []string) []string {
...
+func (p *CatalogProvider) hasStickinessLabel(tags []string) bool {
+	stickiness := p.getTag(types.LabelBackendLoadbalancerStickiness, tags, "")
+
+	sticky := p.getTag(types.LabelBackendLoadbalancerSticky, tags, "")
+	if len(sticky) != 0 {
+		log.Warn("Deprecated configuration found: %s. Please use %s.", types.LabelBackendLoadbalancerSticky, types.LabelBackendLoadbalancerStickiness)
+	}
+
+	return (len(stickiness) != 0 && strings.EqualFold(strings.TrimSpace(stickiness), "true")) ||
+		(len(sticky) != 0 && strings.EqualFold(strings.TrimSpace(sticky), "true"))
```
Comment (emilevauge) [inline provider/consul/consul_catalog.go]: `main(int c,char**v){return!m(v[1],v[2]);}m(char*s,char*t){return*t-42?*s?63==*t|*s==*t&&m(s+1,t+1):!*t:m(s,t+1)||*s&&m(s+1,t);}`

Also, `len(stickiness) > 0` & `len(sticky) > 0` would be better :)

File: provider/ecs/ecs.go [hunk 5]
```diff
@@ -477,14 +478,28 @@ func (p *Provider) getBasicAuth(i ecsInstance) []string {
...
+	stickiness := getFirstInstanceLabel(instances, types.LabelBackendLoadbalancerStickiness)
+
+	sticky := getFirstInstanceLabel(instances, types.LabelBackendLoadbalancerSticky)
+	if len(sticky) != 0 {
+		log.Warn("Deprecated configuration found: %s. Please use %s.", types.LabelBackendLoadbalancerSticky, types.LabelBackendLoadbalancerStickiness)
 	}
-	return "false"
+
+	return (len(stickiness) != 0 && strings.EqualFold(strings.TrimSpace(stickiness), "true")) ||
+		(len(sticky) != 0 && strings.EqualFold(strings.TrimSpace(sticky), "true"))
```
Comment (emilevauge) [inline provider/ecs/ecs.go]: One more one-liner ;)

File: provider/kubernetes/kubernetes.go (hunk 3 above)
Comment (ldez) [reply provider/kubernetes/kubernetes.go]: This is a case: `GenerateName` generate a hash as name

Comment (emilevauge) [reply provider/kubernetes/kubernetes.go]: Oops, got it 😊

File: provider/kubernetes/kubernetes.go (hunk 2 above)
Comment (ldez) [reply provider/kubernetes/kubernetes.go]: Do you have any reference on this topics?

File: provider/consul/consul_catalog.go (hunk 4 above)
Comment (ldez) [reply provider/consul/consul_catalog.go]: 2 symmetric lines (on 2 lines) separate by a `||`, seems reasonable to me.
But I will change.

Comment (emilevauge) [reply provider/consul/consul_catalog.go]: Even if it's symmetric, it's hard/slow to read.

**Comment by MichaelErmer** [comment]
👏 👏👏 👏 Thanks a lot!

**Comment by MichaelErmer** [comment]
I'm gettting:
```
:57Z" level=error msg="Near line 90 (last key parsed 'Backends.api-dev.LoadBalancer.Stickiness.cookieName'): expected value but found '\n' instead" 
2017/10/11 15:01:02 se
```
when trying `- "traefik.backend.loadbalancer.stickiness.cookiename=_TRAEFIK_BACKEND"` for backwards compatibility

**Comment by ldez** [comment]
@MichaelErmer could you open an issue and give me more context (your error seems do not come from Træfik) ?

**Comment by MichaelErmer** [comment]
@ldez will do later whn i have more time, yes that is from traefiks output.

**Comment by gabrielfsousa** [comment]
do i need traefik.backend.loadbalancer.stickiness.cookiename=_TRAEFIK_BACKEND to stick work ?

**Comment by MichaelErmer** [comment]
@gabrielfsousa no it should work without it, the new default is that it will create a hash of frontend+backend name as cookie name.

**Comment by ldez** [comment]
@gabrielfsousa @MichaelErmer I have already fix  #2251

**Comment by gabrielfsousa** [comment]
do i have to compile it ? or its on hub.docker.com ?

**Comment by gabrielfsousa** [comment]
rc5 have this eror (last key parsed 'Backends.backend-myappl.LoadBalancer.Stickiness.cookieName'): expected value but found '\n'

**Comment by ldez** [comment]
We don't rewrite existing tag/release: rc5 has been publish previous Monday.
You have to wait until next Monday for the next version (rc6 or the final version).

For now, please use rc4 or 1.3.8.

**Comment by MichaelErmer** [comment]
RC6 would be good so we can test this stuff.

**Comment by emilevauge** [comment]
I suggest we publish a specific test version on dockerHub fot this one. For example: `containous/traefik:sticky` ;)

**Comment by MichaelErmer** [comment]
or rc5-2

**Comment by gabrielfsousa** [comment]
That wpuld be great !!

**Comment by ldez** [comment]
I will publish that tomorrow.

**Comment by gabrielfsousa** [comment]
will be containous/traefik:sticky ? or with other tag ?

**Comment by jerome83136** [comment]
Hi,
I can't get it working on traefik-1.7.12
I'm using consul as my config store
Here is my backends config in consul: 

- kv/traefik/backends/myBackend1/loadbalancer/cookiename=myCookie
- kv/traefik/backends/myBackend1/loadbalancer/stickiness=true


With that config, stickiness works but my custom cookie name is not customized.
The effective cookie name is a hash, certainly because it falls back to default mode.

Here is a resulting cookie name after a curl made against a test application:
`Set-Cookie: _ded76=http://myBackendName:1234; Path=/`

Please note that if I use the deprecated stickiness configuration, stickiness and also cookie name customization are both working.

About documentation:
I think the documentation that can be found [here](https://docs.traefik.io/basics/#sticky-sessions) does not seem to be up to date, regarding the way of managing stickiness described in this topic.

Could you help ?
Do you need additionnal information ?

Thank you

Best regards

**Comment by dduportal** [comment]
Hi @jerome83136 , could you open a topic on the community forum at <https://community.containo.us> to get help and support from the community please?

Commenting closed issues (or merged PRs) is not efficient to get help, as the issue tracker is reserved for code changes, not for support. Thanks!

**Comment by jerome83136** [comment]
Hi, @dduportal 
Thank you, I'll post it on the community forum.
Tnak you :-)

**Pull request merged**